File: 20.valid-parentheses.0.go
/*
 * @lc app=leetcode id=20 lang=golang
 *
 * [20] Valid Parentheses
 *
 * https://leetcode.com/problems/valid-parentheses/description/
 *
 * algorithms
 * Easy (36.10%)
 * Total Accepted:    546.5K
 * Total Submissions: 1.5M
 * Testcase Example:  '"()"'
 *
 * Given a string containing just the characters '(', ')', '{', '}', '[' and
 * ']', determine if the input string is valid.
 *
 * An input string is valid if:
 *
 *
 * Open brackets must be closed by the same type of brackets.
 * Open brackets must be closed in the correct order.
 *
 *
 * Note that an empty string is also considered valid.
 *
 * Example 1:
 *
 *
 * Input: "()"
 * Output: true
 *
 *
 * Example 2:
 *
 *
 * Input: "()[]{}"
 * Output: true
 *
 *
 * Example 3:
 *
 *
 * Input: "(]"
 * Output: false
 *
 *
 * Example 4:
 *
 *
 * Input: "([)]"
 * Output: false
 *
 *
 * Example 5:
 *
 *
 * Input: "{[]}"
 * Output: true
 *
 *
 */
func isValid(s string) bool {
	expected := []rune{}
	l := 0

	for _, c := range s {
		switch c {
		case '(':
			expected = append(expected, ')')
		case '[':
			expected = append(expected, ']')
		case '{':
			expected = append(expected, '}')
		case ')', ']', '}':
			l = len(expected)
			if l == 0 || expected[l-1] != c {
				return false
			}
			expected = expected[:l-1]
		}
	}

	return len(expected) == 0
}

File: 20.valid-parentheses.0.rs
/*
 * @lc app=leetcode id=20 lang=rust
 *
 * [20] Valid Parentheses
 */
impl Solution {
    pub fn is_valid(s: String) -> bool {
        let mut expected: Vec<char> = vec![];

        for c in s.chars() {
            match c {
                '(' => expected.push(')'),
                '[' => expected.push(']'),
                '{' => expected.push('}'),
                ')' | ']' | '}' => {
                    match expected.pop() {
                        None => {
                            return false
                        }
                        Some(e) if e != c => {
                            return false
                        }
                        _ => {}
                    }
                }
                _ => {}
            }
        }

        return expected.len() == 0
    }
}
